### PR TITLE
feat: Update clean architecture setup repo structure

### PR DESCRIPTION
**Pull request by shreyanshJain852**
## Description

- Add `BaseResponse` for a generic response to application layer
- Handle model validation errors, configured along with base response (for required field, email address, password regex)
- Configure API summary from controller
- Add automapper to all project solution
- Add static file serving middleware, along with http context accessor
- In API program cs
	- configure enum conversions to string for representation in swagger
	- update swagger config to show auth lock only for APIs that are authorized
	- Configure jwt token authentication's life time validator
- Add config values in appsetting for auth token expiry and identity token lifespan
- Add sample application level custom exception for not found
- Reorganise the DTOs in model directory in application layer
- Modify record classes in DTO to have basic data annotations for model validation
- Introduce `BaseEntity` with common fields to be inherited by all DBO entities
- Reorganise enums from classes in domain layer
- Add `Microsoft.EntityFrameworkCore.Tools` package to ensure migration queries are running properly in infrastructure layer
- Reran the migration for updated user entity (with new fields from base entity)
- Move services to a different class library
- Update path for file templates related to email

## Summary by CodeRabbit

- **New Features**
  - Enhanced authentication and user management with structured responses and detailed method summaries.
  - Added authorization policy for admin role.
  - Configured service scope for repositories.
  - Added HTTP client registration.
  - Introduced an enum `ChannelType` with values for different communication channels: SMTP, MailGun, and WhatsApp.
  - Introduced an enum `EmailEvents` with a single value `ForgetPassword`.
- **Refactor**
  - Refactored authentication setup for JWT.
  - Moved enum declarations to a different namespace and updated the enum names.
- **Documentation**
  - Renamed `IEmailRepository` to `IEmailService` for better clarity.
- **Chores**
  - Namespace adjustments, token generation logic updates, and configuration usage modifications.
  - Introduced a new base class `BaseEntity` for common entity properties.

### CODE REVIEW

File: DotnetFoundation/DotnetFoundation.Infrastructure/DependencyInjection.cs [hunk 1]
```diff
@@ -15,46 +15,64 @@
...
+                        // If token expiry time is not null, then validate lifetime
+                        if (expires != null)
+                        {
+                            Validators.ValidateLifetime(notBefore, expires, securityToken, clonedParameters);
+                        }
+
+                        return true;
+                    }
+                };
+            });
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Infrastructure/DependencyInjection.cs]: issuer and audience are validated. JWT key is being stored as env variable

File: DotnetFoundation/DotnetFoundation.Infrastructure/Persistence/UserRepository.cs [hunk 2]
```diff
@@ -44,17 +44,18 @@ public string GenerateJwtToken(UserInfo user)
...
+            _configuration["Jwt:Issuer"],
+            _configuration["Jwt:Audience"],
+            claims,
+            notBefore: DateTime.UtcNow,
+            expires: DateTime.UtcNow.AddHours(Convert.ToDouble(_configuration["Appsettings:LoginAuthTokenExpiryTimeInHrs"])),
+            signingCredentials: credentials
+        );
 
         return new JwtSecurityTokenHandler().WriteToken(token);
     }
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Infrastructure/Persistence/UserRepository.cs]: Jwt key is being fetched from env variable

File: DotnetFoundation/DotnetFoundation.Infrastructure/DependencyInjection.cs [hunk 3]
```diff
@@ -15,46 +15,64 @@ public static class DependencyInjection
 {
     public static IServiceCollection AddInfrastructure(this IServiceCollection services, ConfigurationManager configuration)
     {
+        // Configure database context
         services.AddDbContext<SqlDatabaseContext>(options =>
         {
             string connectionString = configuration.GetConnectionString("DBConnection") ?? throw new Exception("Invalid connection string");
             options.UseMySql(connectionString, ServerVersion.AutoDetect(connectionString));
         });
+
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Infrastructure/DependencyInjection.cs:26]: db related creds are stored as env variables

File: DotnetFoundation/DotnetFoundation.Infrastructure/Persistence/UserRepository.cs [hunk 4]
```diff
@@ -70,13 +71,18 @@
...
                 throw new Exception("Error creating identity user");
             }
+
             IdentityApplicationUser identityApplicationUser = await _userManager.FindByEmailAsync(request.Email).ConfigureAwait(false) ?? throw new Exception("Error finding user");
+
             await AddUserRoleAsync(request.Email, 0).ConfigureAwait(false);
+
             ApplicationUser applicationUser = new()
             {
                 FirstName = request.FirstName,
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Infrastructure/Persistence/UserRepository.cs:89]: updated

File: DotnetFoundation/DotnetFoundation.Application/Exceptions/NotFoundException.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    }
+
+    public NotFoundException(string msg) : base(msg)
+    {
+    }
+
+    public NotFoundException(string message, Exception innerException) : base(message, innerException)
+    {
+    }
+}
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Application/Exceptions/NotFoundException.cs]: updated

File: DotnetFoundation/DotnetFoundation.Domain/Common/BaseEntity.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+namespace DotnetFoundation.Domain.Common;
+public abstract class BaseEntity
+{
+    public int Id { get; set; }
+    public DateTime CreatedOn { get; set; } = DateTime.UtcNow;
+    public int CreatedBy { get; set; }
+    public DateTime ModifiedOn { get; set; }
+    public int ModifiedBy { get; set; }
+    public Status Status { get; set; }
+}
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Domain/Common/BaseEntity.cs]: updated

File: DotnetFoundation/DotnetFoundation.Application/Interfaces/Services/IEmailService.cs [hunk 7]
```diff
@@ -1,8 +1,3 @@
-using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Threading.Tasks;
-
 namespace DotnetFoundation.Application.Interfaces.Services;
 
 /// <summary>
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Application/Interfaces/Services/IEmailService.cs]: updated

File: DotnetFoundation/DotnetFoundation.Application/Models/DTOs/AuthenticationDTO/LoginRequest.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+public record LoginRequest
+{
+    [Required(ErrorMessage = "Email address is required")]
+    [EmailAddress(ErrorMessage = "Invalid email")]
+    public string Email { get; init; }
+
+    [Required(ErrorMessage = "Password is required")]
+    [RegularExpression(@"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,10}$",
+        ErrorMessage = "8-10 character length with 1 uppercase, 1 lowercase, 1 number, 1 special character is required")]
+    public string Password { get; init; }
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Application/Models/DTOs/AuthenticationDTO/LoginRequest.cs]: updated

File: DotnetFoundation/DotnetFoundation.Application/DependencyInjection.cs [hunk 9]
```diff
@@ -1,17 +1,21 @@
...
-        services.AddScoped<IUserService, UserService>();
-        services.AddScoped<IEmailService, EmailService>();
+        // Configure Error Response from Model Validations
+        services.AddMvc().ConfigureApiBehaviorOptions(options =>
+        {
+            options.InvalidModelStateResponseFactory = actionContext =>
+            {
+                return ModelValidationBadRequest.ModelValidationErrorResponse(actionContext);
+            };
+        });
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Application/DependencyInjection.cs:22]: updated

File: DotnetFoundation/DotnetFoundation.Domain/Enums/EmailEvents.cs [hunk 10]
```diff
@@ -0,0 +1,6 @@
+﻿namespace DotnetFoundation.Domain.Enums;
+
+public enum EmailEvents
+{
+    ForgetPassword
+}
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Domain/Enums/EmailEvents.cs]: updated

File: DotnetFoundation/DotnetFoundation.Domain/Enums/ChannelType.cs [hunk 11]
```diff
@@ -0,0 +1,8 @@
+﻿namespace DotnetFoundation.Domain.Enums;
+
+public enum ChannelType
+{
+    SMTP = 1,
+    MailGun = 2,
+    WhatsApp = 3
+}
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Domain/Enums/ChannelType.cs]: updated

File: DotnetFoundation/DotnetFoundation.Infrastructure/DependencyInjection.cs [hunk 12]
```diff
@@ -15,46 +17,64 @@ public static class DependencyInjection
...
+                        // If token expiry time is not null, then validate lifetime
+                        if (expires != null)
+                        {
+                            Validators.ValidateLifetime(notBefore, expires, securityToken, clonedParameters);
+                        }
+
+                        return true;
+                    }
+                };
+            });
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Infrastructure/DependencyInjection.cs]: added explanation for lifetime validator

File: DotnetFoundation/DotnetFoundation.Application/Models/DTOs/AuthenticationDTO/AuthenticationResponse.cs [hunk 13]
```diff
@@ -0,0 +1,5 @@
+namespace DotnetFoundation.Application.Models.DTOs.AuthenticationDTO;
+public record AuthenticationResponse
+{
+    public string Token { get; init; }
+}
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Application/Models/DTOs/AuthenticationDTO/AuthenticationResponse.cs]: updated

File: DotnetFoundation/DotnetFoundation.Domain/Entities/User.cs [hunk 14]
```diff
@@ -1,10 +1,9 @@
...
+using DotnetFoundation.Domain.Common;
 
 namespace DotnetFoundation.Domain.Entities;
 
-public class User
+public class User : BaseEntity
 {
-    public int Id;
-    public string FirstName { get; set; } = null!;
+    public required string FirstName { get; set; }
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Domain/Entities/User.cs]: updated, removed required

File: DotnetFoundation/DotnetFoundation.Application/Models/DTOs/UserDTO/UserRoleRequest.cs [hunk 15]
```diff
@@ -0,0 +1,9 @@
+using DotnetFoundation.Domain.Enums;
+
+namespace DotnetFoundation.Application.Models.DTOs.UserDTO;
+
+public record UserRoleRequest
+{
+    public string Email { get; init; }
+    public Roles Role { get; init; }
+}
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Application/Models/DTOs/UserDTO/UserRoleRequest.cs]: updated

File: DotnetFoundation/DotnetFoundation.Application/Models/DTOs/UserDTO/UserResponse.cs [hunk 16]
```diff
@@ -0,0 +1,8 @@
+namespace DotnetFoundation.Application.Models.DTOs.UserDTO;
+
+public record UserResponse
+{
+    public int Id { get; init; }
+    public string FirstName { get; init; }
+    public string? LastName { get; init; }
+}
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Application/Models/DTOs/UserDTO/UserResponse.cs]: updated

File: DotnetFoundation/DotnetFoundation.Infrastructure/Persistence/EmailRepository.cs [hunk 17]
```diff
@@ -1,10 +1,10 @@
-using System.Net;
-using System.Text;
 using DotnetFoundation.Application.Interfaces.Persistence;
 using DotnetFoundation.Domain.Entities;
+using Microsoft.Extensions.Configuration;
 using Newtonsoft.Json;
 using Newtonsoft.Json.Serialization;
-using Microsoft.Extensions.Configuration;
+using System.Net;
+using System.Text;
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Infrastructure/Persistence/EmailRepository.cs]: using statements formatted using VS ide

File: DotnetFoundation/DotnetFoundation.Infrastructure/DependencyInjection.cs [hunk 18]
```diff
@@ -15,46 +17,65 @@ public static class DependencyInjection
...
-        services.AddScoped<IEmailRepository, EmailRepository>();
+        services.AddScoped<IEmailService, EmailService>();
+
         services.AddHttpClient();
 
         return services;
     }
-}
\ No newline at end of file
+}
```
[the remark this answers is not in the document]
Comment (shreyanshJain852) [reply DotnetFoundation/DotnetFoundation.Infrastructure/DependencyInjection.cs:81]: Http client config is not required to be configured for now, we can do that while implementing projects out of this template

**Review by sameer-s-b — APPROVED** [review]

**Pull request merged**